lib/teleterm/gateway/local_proxy_middleware_test.go

```diff
@@ -0,0 +1,129 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gateway
+
+import (
+	"context"
+	"crypto/tls"
+	"testing"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+
+	"github.com/gravitational/teleport/api/utils/keys"
+	"github.com/gravitational/teleport/lib/defaults"
+	alpn "github.com/gravitational/teleport/lib/srv/alpnproxy"
+	alpncommon "github.com/gravitational/teleport/lib/srv/alpnproxy/common"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+func TestLocalProxyMiddleware_OnNewConnection(t *testing.T) {
+	cert, err := utils.GenerateSelfSignedCert([]string{"localhost"})
+	require.NoError(t, err)
+	tlsCert, err := keys.X509KeyPair(cert.Cert, cert.PrivateKey)
+	require.NoError(t, err)
+	x509cert, err := utils.TLSCertToX509(tlsCert)
+	require.NoError(t, err)
+
+	clockAfterCertExpiry := clockwork.NewFakeClockAt(x509cert.NotAfter)
+	clockAfterCertExpiry.Advance(time.Hour*4 + time.Minute*20)
+
+	clockBeforeCertExpiry := clockwork.NewFakeClockAt(x509cert.NotBefore)
+	clockBeforeCertExpiry.Advance(time.Hour*4 + time.Minute*20)
+
+	validDbRoute := tlsca.RouteToDatabase{
+		Protocol:    defaults.ProtocolPostgres,
+		ServiceName: "foo-database-server",
+	}
+
+	tests := []struct {
+		name        string
+		clock       clockwork.Clock
+		dbRoute     tlsca.RouteToDatabase
+		expectation func(t *testing.T, onNewConnectionErr error, hasCalledOnExpiredCert bool)
+	}{
+		{
+			name:    "With expired cert",
+			clock:   clockAfterCertExpiry,
+			dbRoute: validDbRoute,
+			expectation: func(t *testing.T, onNewConnectionErr error, hasCalledOnExpiredCert bool) {
+				require.NoError(t, onNewConnectionErr)
+				require.True(t, hasCalledOnExpiredCert,
+					"Expected the onExpiredCert callback to be called by the middleware")
+			},
+		},
+		{
+			name:  "With active cert with subject not matching dbRoute",
+			clock: clockBeforeCertExpiry,
+			dbRoute: tlsca.RouteToDatabase{
+				Protocol:    defaults.ProtocolPostgres,
+				ServiceName: "foo-database-server",
+				Username:    "bar",
+				Database:    "quux",
+			},
+			expectation: func(t *testing.T, onNewConnectionErr error, hasCalledOnExpiredCert bool) {
+				require.Error(t, onNewConnectionErr)
+				require.False(t, hasCalledOnExpiredCert,
+					"Expected the onExpiredCert callback to not be called by the middleware")
+			},
+		},
+		{
+			name:    "With valid cert",
+			clock:   clockBeforeCertExpiry,
+			dbRoute: validDbRoute,
+			expectation: func(t *testing.T, onNewConnectionErr error, hasCalledOnExpiredCert bool) {
+				require.NoError(t, onNewConnectionErr)
+				require.False(t, hasCalledOnExpiredCert,
+					"Expected the onExpiredCert callback to not be called by the middleware")
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
```

Comment:
> This is typically done as a way to avoid problems when running the table tests in parallel - did you mean to also add `t.Parallel()` inside `t.Run`?

Ah, no, just a leftover from copying parallel table tests from somewhere else. I don't think it's worth it to execute them in parallel, those are just three very simple tests.